File: test-suite-kotlin/src/test/kotlin/io/micronaut/docs/validation/custom/DurationPatternValidatorSpec.kt
package io.micronaut.docs.validation.custom

import io.micronaut.test.extensions.junit5.annotation.MicronautTest
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.Test
import jakarta.inject.Inject
import jakarta.validation.ConstraintViolationException

@MicronautTest
internal class DurationPatternValidatorSpec {

    // tag::test[]
    @Inject
    lateinit var holidayService: HolidayService

    @Test
    fun testCustomValidator() {
        val exception = assertThrows(ConstraintViolationException::class.java) {
            holidayService.startHoliday("Fred", "junk") // <1>
        }

        assertEquals("startHoliday.duration: invalid duration (junk), additional custom message", exception.message) // <2>
    }
    // end::test[]
}

File: test-suite-python/src/test/python/micronaut/docs/validation/custom/DurationPatternValidatorSpec.py
from typing import Annotated

from jakarta.inject import Inject
from jakarta.validation import ConstraintViolationException
from micronaut.test.extensions.junit5.annotation import MicronautTest
from org.junit.jupiter.api import Test

from .HolidayService import HolidayService


@MicronautTest
class DurationPatternValidatorSpec:

    # tag::test[]
    holiday_service: Annotated[HolidayService, Inject]

    @Test
    def test_custom_validator(self) -> None:
        try:
            self.holiday_service.start_holiday("Fred", "junk")  # <1>
        except ConstraintViolationException as exception:
            assert exception.getMessage() == "start_holiday.duration: invalid duration (junk), additional custom message"  # <2>
        else:
            assert False, "ConstraintViolationException expected"
    # end::test[]
